pkgs/development/compilers/ghc/common-hadrian.nix

```diff
@@ -164,18 +165,33 @@
 , #  Whether to build sphinx documentation.
   # TODO(@sternenseemann): Hadrian ignores the --docs flag if finalStage = Stage1
   enableDocs ? (
-    # Docs disabled if we are building on musl because it's a large task to keep
-    # all `sphinx` dependencies building in this environment.
-    !stdenv.buildPlatform.isMusl
+    # Docs disabled if we are building on musl or cross-building because it's a
+    # large task to keep all `sphinx` dependencies building in this environment.
+    (stdenv.buildPlatform == stdenv.hostPlatform && stdenv.targetPlatform == stdenv.hostPlatform)
+    && !stdenv.buildPlatform.isMusl
   )
 
 , # Whether to disable the large address space allocator
   # necessary fix for iOS: https://www.reddit.com/r/haskell/comments/4ttdz1/building_an_osxi386_to_iosarm64_cross_compiler/d5qvd67/
   disableLargeAddressSpace ? stdenv.targetPlatform.isiOS
+
+, # Whether to build an unregisterised version of GHC.
+  # GHC will normally auto-detect whether it can do a registered build, but this
+  # option will force it to do an unregistered build when set to true.
+  # See https://gitlab.haskell.org/ghc/ghc/-/wikis/building/unregisterised
+  # Registerised RV64 compiler produces programs that segfault
+  # See https://gitlab.haskell.org/ghc/ghc/-/issues/23957
+  enableUnregisterised ? stdenv.hostPlatform.isRiscV64 || stdenv.targetPlatform.isRiscV64
```

Comment:
Registerised riscv64 builds of ghc >=9.6 work for me natively (build = host = target) and this is the default behaviour before this PR. Can you preserve that for all hadrian-based builds?

Reply:
Is that true even when building more complex programs like Pandoc?

From my testing with 9.6.2, registerised initially appeared to work until I tried compiling something more complex.

Reply:
I'll give building with LLVM 15 a try.

In the meantime, I'll undo this change in `common-hadrian.nix`.

Reply:
For older versions of GHC, LLVM 15 is not officially supported, so I don't want to take the risk of fixing registerised builds.

Especially now that 9.6 is the default version.

Reply:
> Is that true even when building more complex programs like Pandoc?
> 
> From my testing with 9.6.2, registerised initially appeared to work until I tried compiling something more complex.

Yes I got pandoc and shellcheck installed and tested basic functionality. Only cachix failed to compile with "error: cycle detected in build of '/nix/store/[..]-cachix-1.7.drv' in the references of output 'bin' from output 'out'" which I have yet to look into. I will try to rebuild a new registerised 9.6.4 from your branch and will report back.

Reply:
9.6.2 still segfaults on my old (~August) NixOS build when using LLVM 15.

```
[ 1 of 16] Compiling Language.Haskell.HsColour.Classify ( Language/Haskell/HsColour/Classify.hs, dist/build/Language/Haskell/HsColour/Classify.o, dist/bu>
/nix/store/mjlk72gj834v6lmbk9j9pvby88s686cz-stdenv-linux/setup: line 1597:   202 Segmentation fault      (core dumped) ./Setup build
```

This was bootstrapped using a native unregisterised 9.2.8, which was itself booted from a cross-compiled 8.10.7 compiled with the original version of this PR. All native packages were compiled for `RV64GC_Zba_Zbb` using `gcc13Stdenv`.

Some (likely useless) kernel logs:

```
ghc_worker[772846]: unhandled signal 11 code 0x1 at 0x0000004e3584b582
CPU: 2 PID: 772846 Comm: ghc_worker Not tainted 6.5.0 #1-NixOS
Hardware name: StarFive VisionFive 2 v1.3B (DT)
epc : 0000004e3584b582 ra : 0000003ff3e06900 sp : 00000004e3ffa730
 gp : 000000000006ea80 tp : 00000004e3fff8e0 t0 : 0000000000004000
 t1 : 0000003ff3e00628 t2 : 0000003ff6f86598 s0 : 00000004f148d652
 s1 : 0000003fedb8afd8 a0 : 0000000000000000 a1 : 0000003fed8dd038
 a2 : 0000000000000002 a3 : 0000000000000002 a4 : 0000000000000032
 a5 : 09e1854e3584b583 a6 : 0000000000000000 a7 : 0000003ff7ffdd50
 s2 : 00000004f311b9f8 s3 : 00000004f16d93c8 s4 : 0000003fed8dc5e0
 s5 : 00000004f31140c0 s6 : 00000004f148d638 s7 : 00000004f36d2870
 s8 : 00000004f2ce1000 s9 : 00000000000003bd s10: 0000003fedb59058
 s11: 00000004f31140c0 t3 : 0000003fed2f0514 t4 : 0000000000000001
 t5 : 0000003fed9c6e2a t6 : 000000000000002e
status: 8000000200006020 badaddr: 0000004e3584b582 cause: 000000000000000c
```

I'll try 9.6.4 on a more up-to-date Nixpkgs to see if the issue still happens.

Reply:
I discovered that the 6.5 kernel has a regression that causes some applications (notably rustc) to segfault. I've upgraded my kernel to 6.6, which doesn't have that issue, but the GHC 9.6.2 I've previously built still segfaults.

(The 9.6.4 build is still ongoing.)

Reply:
The 9.6.4 build is fully working. I have successfully built ShellCheck and nix-output-monitor with it, and not a single segfault has been seen.

Now I need to try rebuilding 9.6.2 to properly rule out the Linux kernel as the cause and try building 9.2 and 9.4 registerised to see if I can get those to work. This is going to take a lot of time.